### DISCUSSION TITLE
Can NodeRED controlled by FlowForge access the file system of the EdgeDevice it is running on?

### DISCUSSION THREAD

**Discussion by Sulomus** [FlowFuse/flowfuse#1520] [Q&A] [answered]
Hello,
the platform looks incredible and we want to experiment with it at our new startup (I can't provide you with a link because it is really new one :D).

### Use case I want to use FlowForge for:

- Our customer has Linux PC's all over the country.
- Each PC is monitoring an industrial process that takes over 4 minutes.
- After the process has ended there is a new .csv file created in a specific file hierarchy.

### We already have the main building blocks:

- NodeRED for data acquisition, InfluxDB for database and webapp for monitoring all the data.
- We want to use FlowForge to control and manage all the NodeRED instances.

### My question is - can NodeRED controlled by FlowForge access the file system of the device it is running on (for example Raspberry PI) and read files / update them?

I am little confused because when I looked into this I found out that [EXEC ](https://flowforge.com/blog/2022/12/node-red-exec-explained/)nodes and [FILES ](https://flowforge.com/blog/2022/11/flowforge-1-1-released/)nodes are disabled.

**Comment by hardillb** [comment, marked as answer]
Hello,

Until recently default Node-RED instances running in the FlowForge managed instance in the cloud have the File and Exec nodes disabled.  This is because when running in the cloud the Node-RED instances run in docker container (in a Kubernetes cluster). This means that there is no persistent filesystem to interact with and no extra commands installed in the container that it would make sense to use with the exec node. The nodes were disabled to prevent users trying to use them in a way that would just lead to failures.

We recently added support for using the File nodes. We have included modified versions of the code Node-RED file nodes in the latest Node-RED Stacks, these are backed by a network service that allows files to be safely stored outside the containers. If you can not see the file nodes in your cloud projects please look under the Projects Settings page to see if you can update the stack to the latest version.

So that covers the instance of Node-RED running in the cloud platform.

Now when snapshots of these projects are deployed using the FlowForge Device Agent (so they run on the remote hardware) the normal core File nodes are loaded so the flows can then interact with the local filesystem on the device.

We do not yet have a solution that will work for the exec node, as mentioned earlier it doesn't make sense in the container environment where there are no commands that could be run, but if there is a strong customer requirement we can look again to see there is a workable solution.

If you would like to setup a call to talk about this please let me know.

**Comment by Sulomus** [comment]
> Now when snapshots of these projects are deployed using the FlowForge Device Agent (so they run on the remote hardware) the normal core File nodes are loaded so the flows can then interact with the local filesystem on the device.

**Q1**: I am little bit confused ... when the snapshots on the EdgeDevices are 1:1 to the NodeRED instance in the cloud - how can I program it to use the default File nodes _(I am specially interested in the `Watch` node because we need to look for changes in the files)_? 

**Q2**: Can I program the FlowForge Device Agent separately from the NodeRED instance in the cloud? 

**Q3**: And if there will be some differences between NodeRED in the FlowForge cloud and the NodeRED on the EdgeDevice how can I track or monitor the EdgeDevices via FlowForge?

I am completely new to this platform and still looking around! Sorry if these question are of the basic kind 😄

**Reply by hardillb** [reply]
A1: The FlowForge file nodes use the same name and have the same interface as the core nodes, you just use the FF version in the cloud and they will automatically be replaced by the core version when the snapshot is deployed to the Device agent. Having said that the Watch node is not supported in the cloud (since nothing other than the Node-RED instance can update the files that are written to the cloud storage).

A2: No, the point is to manage all the device agents from the cloud

A3: There is no difference, you author flows in the cloud and deploy them to the Device Agent.

**Reply by Sulomus** [reply]
A1, A2: Ah okay thank you for clearing that up!
- The watch node can be (to my knowledge) replaced by the filesystem Nodes so I can look into that.

A3: Okay - but there will be difference in the file systems (EdgeDevices will have access to the whole Linux distro but Cloud instance only to the docker files). 

Q1. And then if I run some code in the Cloud I won't have the same output as on the edge device correct?
Q2:  The point I am getting at is how do I debug the EdgeDevices if the file systems are different?

**Reply by hardillb** [reply]
> A3: Okay - but there will be difference in the file systems (EdgeDevices will have access to the whole Linux distro but Cloud instance only to the docker files).

The Cloud instance only have access to a virtual cloud file system, they have NO access to the docker container via the file nodes.

> Q1. And then if I run some code in the Cloud I won't have the same output as on the edge device correct?

It will be very close, and you are free to place files in the same equivalent paths in the cloud filesystem to match the target devices to test.

> Q2: The point I am getting at is how do I debug the EdgeDevices if the file systems are different?

Edge device debugging and testing is an area we continue to iterate on as we get more feedback from users.

The other thing we have is the [Node-RED FlowForge Tools](https://github.com/flowforge/flowforge-nr-tools-plugin)  plugin that allows you to create a flow in the full version of Node-RED and then push that to the cloud and take a snapshot to be deployed to the edge devices. This was new in the last release and again will continue to grow as we iterate on the Edge Device model

**Reply by Sulomus** [reply]
> The Cloud instance only have access to a virtual cloud file system, they have NO access to the docker container via the file nodes.
Ah thanks for clearing that up!

>It will be very close, and you are free to place files in the same equivalent paths in the cloud filesystem to match the target devices to test.
That is true thanks for the advice! 😄 

> The other thing we have is the [Node-RED FlowForge Tools](https://github.com/flowforge/flowforge-nr-tools-plugin) plugin that allows you to create a flow in the full version of Node-RED and then push that to the cloud and take a snapshot to be deployed to the edge devices. This was new in the last release and again will continue to grow as we iterate on the Edge Device model

This look very promising! So I can basically have one main EdgeDevice and push updates from there to the FlowForge Cloud right?

Q1: Can I watch NodeRED logs from EdgeDevices? Or only from the main instance?

**Reply by hardillb** [reply]
> This look very promising! So I can basically have one main EdgeDevice and push updates from there to the FlowForge Cloud right?

Yes,

> Q1: Can I watch NodeRED logs from EdgeDevices? Or only from the main instance?

Only the cloud instance at the moment, as I mentioned we continue to iterate on Edge device debug and testing

**Comment by Sulomus** [comment]
Okay cool! Thank you for your help.

Excited for new features to come! :)
